### ISSUE TITLE
[discussion] How to allocate resources for automl jobs?

### ISSUE THREAD

**Issue by Jeffwan** [ludwig-ai/ludwig#2154]
Seem ludwig has some in-built tools to detect Ray cluster resources and allocate right resources for each trail. 
We plan to create a Ray cluster for each job instead of sharing a large Ray cluster and we run Ray on Kubernetes. However, I am confused how to allocate the total resources for each automl job(ray cluster)? 

Is there a practice or guidance on the resource allocation? For example, evaluate size of the dataset, check number of trails (num_samples) etc and estimate resources range for the job.

**Comment by tgaddair** [comment]
Good question, this is actually something being added in https://github.com/ludwig-ai/ludwig/pull/2134 for distributed training outside of automl. But we plan on applying the same logic to automl as well. (cc @jeffreyftang)

In general, I would say that if you are using GPUs for training, you only really need to think about the GPU resources per trial (we typically set the number of CPUs Ray can see to 0 on GPU workers). Because we use step-based training by default, it's usually sufficient to use 1 GPU per trial in most cases (there is also experimental support for dynamically adding GPUs to trials as other trials are pruned when using hyperband, this is not currently enabled by default in automl, but you can turn it on by editing the `backend` section of the config).

For the total number of GPUs in the cluster, in general the more GPUs you have the more trials you can get through in the time limit, so it's strictly better to have more GPUs if you can afford to in most cases. For Ludwig automl, it's expected that you will likely not be able to get through all samples within the time limit, so the decision is mostly about how much you want to spend / allocate for the job. @amholler might be able to provide more empirical data on how best to set the total number of GPUs per automl job.

If your dataset is large (O(100GB)) I recommend maintaining a separate pool of CPU workers for doing preprocessing and data loading, and using [windowing]() to improve data loading throughput among the different trials. In general, the CPU workers should have a high ratio of RAM to CPUs, and allocate about half the available memory to the Ray object store.

**Comment by amholler** [comment]
TL;DR:
 For tabular & text classification datasets used in AutoML development, 3 simultaneous trials each using 1 T4 GPU, worked well.
 For tabular datasets, hyperparameter search time budgets of 2 hrs were standard.
 For text classification datasets, hyperparameter search time budgets were varied with dataset width, since Bert is quadratic.

Details on AutoML development and validation, including dataset sizes and other aspects of the workloads, are here:
 Tabular AutoML: https://medium.com/predibase/ludwig-automl-for-deep-learning-cf64de9d49c8
 Text Classification AutoML: https://medium.com/predibase/ludwig-automl-for-text-classification-7c1759f3b150

Details on how the resource configuration worked for Tabular AutoML, including trial node usage over time, are here:
 https://www.cncf.io/blog/2022/02/15/managing-public-cloud-resources-for-deep-learning-training-experiments-and-lessons-learned/

**Comment by dalianaliu** [comment]
Resolving this. Please reopen if you have more questions.
